[WIP] JIT Static Hooks: python tests

lobstergrindset opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #49547 [WIP] JIT Static Hooks: cpp tests
* **#49546 [WIP] JIT Static Hooks: python tests**
* #49545 [WIP] JIT Static Hooks: serialization logic
* #49975 [WIP] JIT Static Hooks: schema checking logic
* #49544 [WIP] JIT Static Hooks: adding hooks to class type and adding logic for hook running/compilation

Adds tests to verify that hooks match eager's behavior when created and invoked from python.

Differential Revision: [D25771119](https://our.internmc.facebook.com/intern/diff/D25771119)

Commit:
[WIP] JIT Static Hooks: python tests

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: python tests"




[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: python tests"




[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: python tests"




[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: python tests"




[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: python tests"




[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: python tests"




[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: python tests"




[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: python tests"




[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: python tests"




[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: python tests"




[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: python tests"




[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: python tests"




[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: python tests"




[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: python tests"




[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: python tests"




[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: python tests"


Differential Revision: [D25771119](https://our.internmc.facebook.com/intern/diff/D25771119)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: python tests"


Differential Revision: [D25771119](https://our.internmc.facebook.com/intern/diff/D25771119)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: python tests"


Differential Revision: [D25771119](https://our.internmc.facebook.com/intern/diff/D25771119)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: python tests"


Differential Revision: [D25771119](https://our.internmc.facebook.com/intern/diff/D25771119)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: python tests"


Differential Revision: [D25771119](https://our.internmc.facebook.com/intern/diff/D25771119)

[ghstack-poisoned]

SplitInfinity commented on test/jit/test_hooks.py:
```diff
@@ -0,0 +1,757 @@
+
+class OuterModuleSingleIO(torch.nn.Module):
+    def __init__(self, name: str, submodule_name: str):
+        super(OuterModuleSingleIO, self).__init__()
```
```suggestion
        super().__init__()
```

SplitInfinity commented on test/jit/test_hooks.py:
```diff
@@ -0,0 +1,757 @@
+        self.name = name
+
+    def forward(self, input: Tuple[str, str]):
+        input_access1 = input[0]
+        input_access2 = input[1]
```
```suggestion
        input_access1, input_access2 = input
```

SplitInfinity commented on test/jit/test_hooks.py:
```diff
@@ -0,0 +1,757 @@
+        self.submodule = InnerModuleTupleDoubleIO(submodule_name)
+
+    def forward(self, input: Tuple[str, str]):
+        input_access1 = input[0]
+        input_access2 = input[1]
```
```suggestion
        input_access1, input_access2 = input
```

SplitInfinity commented on test/jit/test_hooks.py:
```diff
@@ -0,0 +1,757 @@
+        def pre_hook(self, input: Tuple[List[str], str]):
+            assert self.name == "outer_mod_name"
+            assert input[0][0] == "a"
+            return ["pre_hook_overrid_name"], "pre_hook_override"
```
```suggestion
            return ["pre_hook_override_name"], "pre_hook_override"
```

SplitInfinity commented on test/jit/test_hooks.py:
```diff
@@ -0,0 +1,757 @@
+        ):
+            assert self.name == "outer_mod_name"
+            assert (
+                input[0][0] == "pre_hook_overrid_name"
```
```suggestion
                input[0][0] == "pre_hook_override_name"
```

SplitInfinity commented on test/jit/test_hooks.py:
```diff
@@ -0,0 +1,757 @@
+            assert self.name == "outer_mod_name"
+            assert (
+                input[0][0] == "pre_hook_overrid_name"
+            )  # what the pre_hook overrid instead of the original
```
Can you clarify this comment?

SplitInfinity commented on test/jit/test_hooks.py:
```diff
@@ -0,0 +1,757 @@
+        m.register_forward_hook(forward_hook)
+        m.register_forward_hook(forward_hook)
+
+        scripted = torch.jit.script(m)
```
Is this needed?
```suggestion
```

SplitInfinity commented on test/jit/test_hooks.py:
```diff
@@ -0,0 +1,757 @@
+
+        err_msg = (
+            "Can't define hook: foo on class: .+ "
+            "because a method or hook with that name already exists."
```
So this tests collisions between hooks and methods, but what about hooks and other hooks? I guess redefining a hook function is tricky, but something like this might work:
```
def hook_fn(...):
   ...

mod.register_forward_pre_hook(hook_fn)

def hook_fn(...):
   ...
```
I know the interpreter accepts a redefinition like this, but I'm not sure about redefinitions in source.

Lilyjjo replied:
Good catch, the existing code allowed this and was running the first hook definition twice. I changed the logic in _recursive.py 'get_hook_stubs' on the bottom commit to check for this. I also added the test case to check for it.

SplitInfinity reviewed (commented):
Could you add more comments explaining what's being tested in each of the test cases? Similar to how you have 
```
        # Tests if hook error message is printed out if erroring before schema check.
        # Useful for when user is scripting hooks while not aware of it
```
for one of them. It's difficult (for me at least) to figure out what's going on based on the names of all the test modules and some of the test cases.

The tests for the error cases look thorough though, and I saw tests for the basic things I had in mind:
- same hook used twice
- output override
- string of forward hooks where one of them changes the output type
- tuple-ing of the output of prehooks if it is not already a tuple
- direct `forward` invocation should not call hooks hooks
- hooks on a parent module and its submodule
- hook + method name collision
- hooks can access module internal state

One thing I'm curious about: what does the error message look like when a hook encounters a `RuntimeError`?

Commit:
Update on "[WIP] JIT Static Hooks: python tests"


Differential Revision: [D25771119](https://our.internmc.facebook.com/intern/diff/D25771119)

[ghstack-poisoned]

Lilyjjo commented:
I renamed all of the modules and tests, I agree that they were hard to decipher. I also added comments

Commit:
Update on "[WIP] JIT Static Hooks: python tests"


Differential Revision: [D25771119](https://our.internmc.facebook.com/intern/diff/D25771119)

[ghstack-poisoned]

Lilyjjo commented:
Moved shared code for cpp and python tests into new file

Commit:
Update on "[WIP] JIT Static Hooks: python tests"


Differential Revision: [D25771119](https://our.internmc.facebook.com/intern/diff/D25771119)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: python tests"


Differential Revision: [D25771119](https://our.internmc.facebook.com/intern/diff/D25771119)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: python tests"


Differential Revision: [D25771119](https://our.internmc.facebook.com/intern/diff/D25771119)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: python tests"


Differential Revision: [D25771119](https://our.internmc.facebook.com/intern/diff/D25771119)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: python tests"


Differential Revision: [D25771119](https://our.internmc.facebook.com/intern/diff/D25771119)

[ghstack-poisoned]

SplitInfinity commented on test/jit/test_hooks_modules.py:
```diff
@@ -0,0 +1,531 @@
+        return self.submodule((1,))
+
+
+# Tests for JIT forward hooks and pre-hooks
```
Nit: I would clarify that these are not tests in and of themselves; they set up modules with hooks attached that are used elsewhere in Python and C++ tests alike. I would update the comments below in a similar fashion.

SplitInfinity commented on test/jit/test_hooks_modules.py:
```diff
@@ -0,0 +1,531 @@
+
+
+# Tests for JIT forward hooks and pre-hooks
+def test_module_no_forward_input_model():
```
Nit: Usually `test_` is used for test methods of a class that extends `unittest.TestCase` and it might confuse people to see such a prefix on a function that is not a unit test. Consider renaming these to something like `create_module_no_forward_input()`.

SplitInfinity commented on test/jit/test_hooks_modules.py:
```diff
@@ -0,0 +1,531 @@
+
+
+def test_module_hook_return_nothing_model():
+    # Test module level hooks that reutrn nothing
```
```suggestion
    # Test module level hooks that return nothing
```

SplitInfinity commented on test/jit/test_hooks_modules.py:
```diff
@@ -0,0 +1,531 @@
+    # in a tuple when the returned pre-hook result isn't a tuple
+    # (to allow the result to be passed to another pre-hook if needed).
+    # The eager behavior doesn't wrap the single tuple input pre-hook return in a
+    # tuple as it should. To get consitent behavior between single tuple inputs and
```
```suggestion
    # tuple as it should. To get consistent behavior between single tuple inputs and
```

SplitInfinity commented on test/jit/test_hooks_modules.py:
```diff
@@ -0,0 +1,531 @@
+    return m
+
+
+def test_forward_tuple_input_model():
```
This is a good catch!

SplitInfinity commented on test/jit/test_hooks_modules.py:
```diff
@@ -0,0 +1,531 @@
+    m_submod_call.submodule.register_forward_pre_hook(pre_hook)
+    m_submod_call.submodule.register_forward_hook(forward_hook)
+
+    m_submod_forward_call_scripted = torch.jit.script(m_submod_forward_call)
+    m_submod_call_scripted = torch.jit.script(m_submod_call)
+
+    return m
```
Where is `m` defined? Why are the scripted versions of `m_submod_forward_call` and  `m_submod_call` not used?

SplitInfinity replied:
I also noticed that this test is absent from the C++ tests stacked on top. Why is that?

Lilyjjo replied:
It's actually not used anywhere, I think it got made somehow in the code refactoring at some point. I'll just delete it

SplitInfinity commented on test/jit/test_hooks.py:
```diff
@@ -0,0 +1,357 @@
+            m_submod_forward_call_scripted("a"), m_submod_call_scripted("a")
+        )
+
+    """ TODO: add this test back once figured out how to print 
```
Nit: I would prefer to use `@unittest.skip` here instead of commenting the test out like this but it's just personal preference.

SplitInfinity commented on test/jit/test_hooks.py:
```diff
@@ -0,0 +1,357 @@
+
+
+# Tests for JIT forward hooks and pre-hooks
+class TestHooks(JitTestCase):
```
One more test case I would add (if it's not there already) is accessing a submodule, storing it in a temporary variable, and then calling it:
```
m = torch.jit.script(module_with_hooks)
sm = m.submod_with_hooks
output = sm(input)
# check that output is ok, hooks should have been called for `sm`
```

SplitInfinity replied:
And then we also need a C++ version of this.

SplitInfinity reviewed (commented):
Nice! Looks much cleaner now.

Commit:
Update on "[WIP] JIT Static Hooks: python tests"


Differential Revision: [D25771119](https://our.internmc.facebook.com/intern/diff/D25771119)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: python tests"


Differential Revision: [D25771119](https://our.internmc.facebook.com/intern/diff/D25771119)

[ghstack-poisoned]

SplitInfinity commented on test/jit/test_hooks.py:
```diff
@@ -158,6 +179,7 @@ def forward_hook(self, input: Tuple[str], output: str):
         self.assertEqual(m.forward("a"), m_scripted.forward("a"))
         self.assertNotEqual(m_scripted("a"), m_scripted.forward("a"))
 
+    @unittest.skipIf(IS_SANDCASTLE or IS_FBCODE, saving_msg)
```
As per Python documentation, you can skip the entire class using the same decorator instead of each method separately:
```
@unittest.skip("showing class skipping")
class MySkippedTestCase(unittest.TestCase):
    def test_not_run(self):
        pass
```

Lilyjjo replied:
There are a few tests that can run still, the hook error messages. I wasn't sure if I should skip those or not

SplitInfinity replied:
Oops, forgot about those.

Commit:
Update on "[WIP] JIT Static Hooks: python tests"


Differential Revision: [D25771119](https://our.internmc.facebook.com/intern/diff/D25771119)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: python tests"


Differential Revision: [D25771119](https://our.internmc.facebook.com/intern/diff/D25771119)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: python tests"


Adds tests to verify that hooks match eager's behavior when created and invoked from python.

Differential Revision: [D25771119](https://our.internmc.facebook.com/intern/diff/D25771119)

[ghstack-poisoned]

SplitInfinity reviewed (approved):
Looks good! Let's make sure the rest of the CI tests that have not finished at the time that I am writing this are okay.

SplitInfinity commented:
Another follow up: a test with a more complicated hook that closes over a variable in its outer scope (basically to make sure we're constructing the resolution callback correctly).

Lilyjjo commented:
> Another follow up: a test with a more complicated hook that closes over a variable in its outer scope (basically to make sure we're constructing the resolution callback correctly).

Do you know what the expected behavior is? 
```
def test_resolution_callback_closure(self):
        # Test that hooks can close over variables from their
        # outer python scopes
        m = ModuleForwardSingleInput("outer_mod_name", "inner_mod_name")

        a = "pre_hook_override_name"

        def pre_hook(self, input: Tuple[str]) -> Tuple[str]:
            return (a,)

        def forward_hook(self, input: Tuple[str], output: str):
            assert self.name == "outer_mod_name"
            assert input == (a,)
            output = output + "_fh"
            return output

        m.register_forward_pre_hook(pre_hook)
        m.register_forward_hook(forward_hook)

        m("a")

        m_scripted = torch.jit.script(m)

        #self.checkModule(m, ("a"))
```
This produces this error: `RuntimeError: 
python value of type 'str' cannot be used as a value. Perhaps it is a closed over global variable? If so, please consider passing it in as an argument or use a local varible instead.`

I'm not sure how closures are expected to work

SplitInfinity commented:
Sorry I should have been more specific - can you alias a type in the outer scope and use that in a hook?
```
def test_resolution_callback_closure(self):
        # Test that hooks can close over variables from their
        # outer python scopes
        m = ModuleForwardSingleInput("outer_mod_name", "inner_mod_name")

        _string = str

        def pre_hook(self, input: Tuple[_string]) -> Tuple[_string]:
           ...

        def forward_hook(self, input: Tuple[_string], output: _string):
            ...

        m.register_forward_pre_hook(pre_hook)
        m.register_forward_hook(forward_hook)

        m("a")

        m_scripted = torch.jit.script(m)

        #self.checkModule(m, ("a"))
```

Commit:
Update on "[WIP] JIT Static Hooks: python tests"


Adds tests to verify that hooks match eager's behavior when created and invoked from python.

Differential Revision: [D25771119](https://our.internmc.facebook.com/intern/diff/D25771119)

[ghstack-poisoned]

Lilyjjo commented:
Yeah that still gives an error: 
```
File "/home/lillianjohnson/local/miniconda3/envs/pytorch/lib/python3.8/site-packages/torch/jit/_recursive.py", line 337, in create_hooks_from_stubs
    concrete_type._create_hooks(hook_defs, hook_rcbs, pre_hook_defs, pre_hook_rcbs)
RuntimeError: 
Unknown type name '_string':
  File "/home/lillianjohnson/local/pytorch/test/jit/test_hooks.py", line 121
        def forward_hook(self, input: Tuple[_string], output: _string):
                                            ~~~~~~~ <--- HERE
            assert self.name == "outer_mod_name"
            assert input == ("pre_hook_override_name",)
```

SplitInfinity commented:
Okay, let's look into and handle that in a follow-up PR. It would be good to land this stack to avoid having to rebase and deal with merge conflicts. There's still a lot of progress that is worth committing.

Lilyjjo commented:
I used the same function to make the hook rcbs that is used to make the method rcbs: 
```
def make_stub(func, name):
    rcb = _jit_internal.createResolutionCallbackFromClosure(func)
    ast = get_jit_def(func, name, self_name="RecursiveScriptModule")
    return ScriptMethodStub(rcb, ast, func)
```
I'm not totally sure how to get this to work, but I can look into it

Commit:
Update on "[WIP] JIT Static Hooks: python tests"


Adds tests to verify that hooks match eager's behavior when created and invoked from python.

Differential Revision: [D25771119](https://our.internmc.facebook.com/intern/diff/D25771119)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: python tests"


Adds tests to verify that hooks match eager's behavior when created and invoked from python.

Differential Revision: [D25771119](https://our.internmc.facebook.com/intern/diff/D25771119)

[ghstack-poisoned]
